### PR TITLE
Fix initializing the global hash map with a small stack

### PR DESCRIPTION
**Pull request by SimonSapin**
Fix https://github.com/servo/html5ever/issues/393

---
This change is [Reviewable](https://reviewable.io/reviews/servo/string-cache/227)

### CODE REVIEW

File: src/atom.rs [hunk 1]
```diff
@@ -67,8 +67,13 @@ impl StringCacheEntry {
 
 impl StringCache {
     fn new() -> StringCache {
+        type T = Option<Box<StringCacheEntry>>;
+        assert!(std::mem::size_of::<T>() == std::mem::size_of::<usize>());
```
Comment (nox) [inline src/atom.rs]: This can be checked statically with `std::mem::transmute`.

File: src/atom.rs [hunk 2]
```diff
@@ -67,8 +67,13 @@ impl StringCacheEntry {
...
+        let vec = std::mem::ManuallyDrop::new(vec![0_usize; NB_BUCKETS]);
         StringCache {
-            buckets: unsafe { mem::zeroed() },
+            buckets: unsafe {
+                Box::from_raw(vec.as_ptr() as *mut [T; NB_BUCKETS])
```
Comment (nox) [inline src/atom.rs]: Why not just `vec![None; NB_BUCKETS].into_boxed_slice()` anyway?

Comment (SimonSapin) [reply src/atom.rs]: The bound check when indexing with a masked hash can likely be optimized away if the field is a fixed-size array, maybe not if it’s a slice. Though I haven’t verified either.

https://github.com/servo/string-cache/blob/51b58ce6f94c2b245410bfd70c5d6e1fb7b2a7e0/src/atom.rs#L76-L79

Comment (nox) [reply src/atom.rs]: Oh I didn't notice the array. My issue wasn't with that part, but with the `0usize` becoming `None` and the size check, I don't think we need that. You can just do `vec![None; NB_BUCKETS]` to create `vec`.

Comment (SimonSapin) [reply src/atom.rs]: I tried `vec![None; NB_BUCKETS]`, it requires `StringCacheEntry: Clone` and deriving that doesn’t work because `AtomicIsize: !Clone`

Comment (nox) [reply src/atom.rs]: Oh I see!

Comment (SimonSapin) [reply src/atom.rs]: (I even made https://github.com/rust-lang/rust/pull/64891 in anticipation of this…)

File: src/atom.rs (hunk 1 above)
Comment (SimonSapin) [reply src/atom.rs]: Done

**Pull request merged**